src/ngraph/op/experimental/shape_of.hpp

```diff
@@ -39,6 +39,17 @@ namespace ngraph
                     copy_with_new_args(const NodeVector& new_args) const override;
 
                 void validate_and_infer_types() override;
+
+                // FOR CONSTANT FOLDING INTERNAL USAGE ONLY
+                // Constant folding for cases with static rank but dynamic shape create a subgraph
+                // which contains a Shape of.
+                // In this case we need to prevent constant folding from endless creation of these
+                // subgraphs.
+                // These metods should be removed if better solution will be designed.
+                void block_recurrent_constant_folding() { m_is_foldable = false; }
+                bool is_foldable() { return m_is_foldable; }
```

Comment:
Alternate solution proposal: Cache the sub-graph in the op instance and have constant-folding use the cached value if  it there.

Reply:
Caching the sub-graph in the op instance could make an unnecessary overhaul if model would have more than one `shape_of`.

I recreated these two methods as getter and setter.
I also preserve this attribute in the cloning operation.